### ISSUE TITLE
Feature Request: Find and replace instances with UUID from popup

### ISSUE THREAD

**Issue by Alacho2** [leomillon/uuid-generator-plugin#46]
I love this plugin, it's excellent and covered exactly what I needed.
But I would love to see a way, from the popup window, of replacing instances for text with an idea, instead of just copying to clip board.

Let's say you generate a long list of GUID's from the popup. Hundreds or even thousands, to be used in an existing structure of objects. With webstorm's built-in functionality, I can easily manage to create an extra property on the object (javascript). But I still have to manually go through each of those properties, assigning a GUID.

What I'm proposing is functionality for "Find and Replace" from the popup window. You generate 400 GUID's, write a text (or regex pattern), click replace and the plugin goes through those instances, replacing it with the generated GUID.

I hope this makes sense.

**Comment by leomillon** [comment]
Hi @Alacho2 !

Ok I think I mostly understand your need except regarding the "existing structure of objects" part.

Could you please give me an example of code (before vs. after) to illustrate your use case?

On the other hand, if you "simply" want to replace existing UUIDs from an existing code, have you tried this recently added "find and replace distinct" new feature?
You can have a preview here: https://github.com/leomillon/uuid-generator-plugin/issues/41#issuecomment-791968918

**Comment by Alacho2** [comment]
That's a neat one, but doesn't exactly cover the use-case. I guess it's kinda an edge-case what I'm asking for.

Consider the following array in JavaScript: 

```
const values = [{
  name: "Some name",
  category: "CATEGORY_OF_RELATED",
  action: () => {}
  label: "An action",
  parameter: "55",
},{
  name: "Some name",
  category: "CATEGORY_OF_RELATED",
  action: () => {}
  label: "An action",
  parameter: "55",
},{
  name: "Some name",
  category: "CATEGORY_OF_RELATED",
  action: () => {}
  label: "An action",
  parameter: "55",
},{
  name: "Some name",
  category: "CATEGORY_OF_RELATED",
  action: () => {}
  label: "An action",
  parameter: "55",
},{
  name: "Some name",
  category: "CATEGORY_OF_RELATED",
  action: () => {}
  label: "An action",
  parameter: "55",
},{
  name: "Some name",
  category: "CATEGORY_OF_RELATED",
  action: () => {}
  label: "An action",
  parameter: "55",
}, /* the 396 others are ommitted */]
```

Neither of these objects has a unique identifier. 
Now, I could easily use Webstorm's Find and Replace functionality to add a GUID property, so an object would look like this: 
```
{
  name: "Some name",
  category: "CATEGORY_OF_RELATED",
  action: () => {}
  label: "An action",
  parameter: "55",
  guid: '-UUID-'
}
```

What I'm asking for is your plugin to be able to find the text '-UUID-' (or any other example), and replace it with the GUID generated by the popup window. 

So the object would look like 
```
{
  name: "Some name",
  category: "CATEGORY_OF_RELATED",
  action: () => {}
  label: "An action",
  parameter: "55",
  guid: 'adfc6112-89a8-4212-ab51-e619cd9f6e32'
}
```

**Comment by leomillon** [comment]
Ok I see your point.

As long as you can "prepare the field" with target placeholders I think that we can work out something to suit your use case.

I'll try to think about a flexible solution to add a "dynamic templating" feature with the following goals to reach:

- A specific placeholder for the plugin to avoid potential conflicts with other existing placeholders from various templating systems (ex: `$gen.***$`)
- A typed placeholder to target the ID type you want to generate between `UUID`/`ULID`/`CLID` (ex: `$gen.uuid$`)
- Maybe add some kind of "classifier" to give the intention of the target replacement: completely random or "distinct" (ex: `$gen.uuid$` will always be a new random one vs. `$gen.uuid.label_1$` will be replaced by the same generated UUID everywhere it has been found)

What do you think?

**Comment by Alacho2** [comment]
That sounds absolutely awesome and perfect! As long as it is as simple as you've currently made the popup, I'll be very happy. :)

**Comment by leomillon** [comment]
The problem with the existing popup system is that it does not define the "code scope" to target the replacement :/

So there are 2 things I've got in mind:
1. Re-use the "find and replace distinct..." system (meaning that you select and use the dedicated action to apply the replacement without configuration)
2. If possible, I could create a variant of the previous point to allow "some configuration" regarding the replacement strategy but that could be tricky

Does the first point could just suit your need? What would you expect to find in the "popup" (point 2.)?

**Comment by Alacho2** [comment]
The first point would be very suiting, indeed. That is what I tried to use, originally, when first doing this task.

**Comment by leomillon** [comment]
Hi @Alacho2 

Here is what I've managed to produce:

- ID placeholders for all supported ID types so far (`#gen.uuid#`, `#gen.ulid#`, `#gen.cuid#`)
- Each ID placeholder can have a "label" to register a "reused" generated value for the same labeled placeholder in the same selection
- I added some simple highlighting to give the user some feedback about a properly written placeholder
- I also added a quick fix on the placeholder to replace this value (and only this one)

Here is a preview of the result:

- ID placeholders highlight

<img width="904" alt="id_placeholders_highlight" src="https://user-images.githubusercontent.com/1232793/116727777-4b3fa800-a9e5-11eb-8a34-5ae0cd4712a6.png">

- ID placeholders replacement action

![id_placeholders_replacement](https://user-images.githubusercontent.com/1232793/116727852-64485900-a9e5-11eb-9178-78412f2d1a75.gif)

Feel free to give your feedback before I merge this PR :)

**Comment by leomillon** [comment]
@Alacho2 this feature is now available in the last 4.4.0 release : https://plugins.jetbrains.com/plugin/8320-uuid-generator
